### PR TITLE
[streamcz] fix extractor

### PR DESCRIPTION
**Pull request by arkamar**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This is a new extractor I developed quite some time ago for televizeseznam.cz and submitted to https://github.com/ytdl-org/youtube-dl/pull/21774. I have recently found yt-dlp and let's hope I'll be luckier here :) I rebased the original branch on top of a recent `master` branch, and did few modifications:
1. update `_VALID_URL` to support `stream.cz` as well,
2. remove old broken `StreamCZ` extractor in favor of this one,
3. add one more test with `url` pointing to `stream.cz`.

Closes: #1329
Fixes: #1731

### CODE REVIEW

**Comment by mkubecek** [comment]
Works fine for me for www.stream.cz and legacy www.televizeseznam.cz URLs which get redirected to www.stream.cz (https://www.televizeseznam.cz/video/...) but not for current televizeseznam.cz content (live streaming and archive of Seznam TV), e.g. https://www.televizeseznam.cz/tv?tf=1636498800&dotId=4795782  These seem to use a different format, I tried to look into it a bit but I'm not very familiar with how all this works.

If you are interested, I have a minor improvement in https://github.com/mkubecek/yt-dlp/commit/f0faa78c7859

**Comment by arkamar** [comment]
@mkubecek thanks, I have added your change to this branch. Anyway, since this extractor supports www.stream.cz and only legacy www.televizeseznam.cz URLs it seems to be appropriate to rename it to `StreamCZ` which I just did in last commit.

**Comment by arkamar** [comment]
Well, just to clarify this change set a little bit. The `stream.cz` was rebranded to `televizeseznam.cz` few years back (2018 I think) and they changed the API as well. I developed `TelevizeSeznam` extractor in middle of 2019. However, part of `televizeseznam.cz` was rebranded back to `stream.cz` later (this year, 2021) and this is why I renamed original `TelevizeSeznam` extractor to `StreamCZ`, which fixes currently present `StreamCZ` extractor and it supports both, legacy `televizeseznam.cz` and current `stream.cz` urls.

File: yt_dlp/extractor/streamcz.py [hunk 1]
```diff
@@ -1,105 +1,125 @@
...
-        'url': 'http://www.stream.cz/peklonataliri/765767-ecka-pro-deti',
-        'md5': '934bb6a6d220d99c010783c9719960d5',
+        'url': 'https://www.televizeseznam.cz/video/lajna/buh-57953890',
+        'md5': '40c41ade1464a390a0b447e333df4239',
         'info_dict': {
-            'id': '765767',
+            'id': '57953890',
+            'display_id': 'buh',
+            'title': 'Bůh',
+            'description': 'Trenér Hrouzek je plný rozporů. Na pomoc si povolá i toho nejvyššího. Kdo to ale je? Pomůže mu vyřešit několik dilemat, která se mu v poslední době v životě nahromadila?',
```
Comment (Ashish0804) [inline yt_dlp/extractor/streamcz.py:35]: md5 this, do same for the description below

File: yt_dlp/extractor/streamcz.py [hunk 2]
```diff
@@ -1,105 +1,125 @@
...
+        if not play_list:
+            return None
+
+        subtitles = {}
+        for k, v in play_list.items():
+            if v.get('language'):
+                for ext in v.get('urls'):
+                    subtitles.setdefault(v['language'], []).append({
+                        'ext': ext,
+                        'url': urljoin(spl_url, v['urls'].get(ext))
```
Comment (Ashish0804) [inline yt_dlp/extractor/streamcz.py:61]: i don't like this tbh. Maybe check for it above append only if it exists

File: yt_dlp/extractor/streamcz.py [hunk 3]
```diff
@@ -1,105 +1,125 @@
...
-                    'format_note': '%s-%s' % (qlabel, typ) if qlabel else typ,
-                    'format_id': '%s-%s' % (typ, f['quality']),
-                    'url': f['source'],
-                    'height': int_or_none(f['quality'].rstrip('p')),
-                    'quality': quality,
-                })
+        for r, v in play_list.items():
+            format = {
+                'format_id': r,
+                'url': urljoin(spl_url, v.get('url')),
```
Comment (Ashish0804) [inline yt_dlp/extractor/streamcz.py:70]: if this doesn't exist then just continue

File: yt_dlp/extractor/streamcz.py [hunk 4]
```diff
@@ -1,105 +1,125 @@
...
+                'protocol': 'https',
+                'ext': ext
+            }
+            if v.get('bandwidth'):
+                format.update({'tbr': v['bandwidth'] / 1000})
+                if v.get('duration'):
+                    format.update({'filesize_approx': v['bandwidth'] * v['duration'] / 8000})
```
Comment (Ashish0804) [inline yt_dlp/extractor/streamcz.py:77]: Don't think u have to do this with the new commits.
Just set the duration

File: yt_dlp/extractor/streamcz.py [hunk 5]
```diff
@@ -1,105 +1,125 @@
...
+            formats.extend(self._extract(None, spl_url, play_list['http_stream']['qualities']))
+
+        if play_list.get('mp4'):
+            formats.extend(self._extract('mp4', spl_url, play_list['mp4']))
+
         self._sort_formats(formats)
+        return formats
+
+    def _real_extract(self, url):
+        display_id, video_id = re.match(self._VALID_URL, url).groups()
```
Comment (Ashish0804) [inline yt_dlp/extractor/streamcz.py:99]:
```suggestion
        display_id, video_id = self._match_valid_url(url).groups()
```

**Comment by arkamar** [comment]
I addressed all mentioned isses from review.

**Comment by arkamar** [comment]
ping

**Comment by xmstspider** [comment]
Tested [this](https://www.stream.cz/tajemno/znicehonic-jim-skrz-strechu-prolitnul-zahadny-predmet-badatele-vse-objasnili-64147267) URL and still working. Using version 2021.12.17.

**Comment by arkamar** [comment]
@xmstspider thanks.

**Comment by xmstspider** [comment]
> @xmstspider thanks.

Sorry I made I typo in my previous comment, which I edited now. The latest version (2021.12.17) still tries to use the generic extractor, while mkubecek's version picks up the streamcz.

**Comment by arkamar** [comment]
@xmstspider yes, that makes sense. The latest version still uses generic extractor because this PR is not merged yet. Anyway, this PR is based on https://github.com/arkamar/yt-dlp/tree/televizeseznam which shares most of the history with mkubecek's one plus few more changes.

File: yt_dlp/extractor/streamcz.py [hunk 6]
```diff
@@ -1,105 +1,105 @@
...
             'id': video_id,
-            'title': title,
-            'thumbnail': thumbnail,
+            'display_id': display_id,
+            'title': data.get('name'),
+            'description': data.get('perex'),
             'formats': formats,
-            'description': data.get('web_site_text'),
-            'duration': int_or_none(data.get('duration')),
-            'view_count': int_or_none(data.get('views')),
```
Comment (pukkandan) [inline yt_dlp/extractor/streamcz.py:103]: are the rest of these metadata no longer extractable?

File: yt_dlp/extractor/streamcz.py [hunk 7]
```diff
@@ -1,105 +1,104 @@
...
+        video = metadata['data']
 
         subtitles = {}
-        srt_url = data.get('subtitles_srt')
-        if srt_url:
-            subtitles['cs'] = [{
-                'ext': 'srt',
-                'url': srt_url,
-            }]
+        for subs in video.get('subtitles').values() or []:
```
Comment (arkamar) [inline yt_dlp/extractor/streamcz.py:85]: This is not correct, `video` does not have to contain `subtitles` in which case exception is thrown. I'll fix it.

**Pull request merged**